detox/src/devices/common/drivers/android/tools/HashHelper.js

```diff
@@ -0,0 +1,46 @@
+const crypto = require('crypto');
+const { DetoxRuntimeError } = require('../../../../../errors');
+
+const SUPPORTED_HASH_TYPES = {
+  MD5: 'md5',
+}
+
+const HASH_PATH = '/data/local/tmp/detox';
+
+class HashHelper {
+  constructor(adb, hashXfer, hashPath = HASH_PATH) {
+    this._adb = adb;
+    this._hashXfer  = hashXfer;
+    this._hashPath = hashPath;
+  }
+
+  generateHash(path, hashType = SUPPORTED_HASH_TYPES.MD5) {
+    if (!path) {
+      throw new DetoxRuntimeError({
+        message: `Path must be provided for hash generation`,
+      });
+    }
+
+    switch (hashType) {
+      case SUPPORTED_HASH_TYPES.MD5:
+        return crypto.createHash('md5').update(path).digest("hex");
+      default:
+        throw new DetoxRuntimeError({
+        message: `Hashtype is unsupported: ${hashType}`,
+        hint: `Use a supported hashtype, such as MD5`,
+      });
+    }
+  }
+
+  async saveHashToRemote(deviceId, bundleId, hash) {
+    const hashFileName = `${bundleId}.hash`;
+    await this._adb.createFileWithContent(deviceId, this._hashPath, hashFileName, hash);
+  }
+
+  async compareRemoteToLocal(deviceId, bundleId, localHash) {
+    const remoteHash = await this._hashXfer.readHashFile(deviceId, bundleId);
```

Comment:
What if the user:
1. runs Detox tests
2. re-installs a different version of the app on the device using React Native CLI or some other manual way, outside of Detox flows
3. returns to the initial project and runs Detox tests again?

Won't Detox treat it as if the hash did not change (hence, no need to reinstall), while it should have been more paranoid?

Reply:
Actually won't it be wiser to check last install timestamp or maybe use those hashes from dumpsys:

```
adb shell dumpsys package com.example.testapp.first

Packages:
  Package [com.example.testapp.first] (2036fd1):
    userId=10225
    pkg=Package{42e7a36 com.example.testapp.first}
    codePath=/data/app/com.example.testapp.first-1
    resourcePath=/data/app/com.example.testapp.first-1
    legacyNativeLibraryDir=/data/app/com.example.testapp.first-1/lib
    primaryCpuAbi=null
    secondaryCpuAbi=null
    versionCode=1 minSdk=21 targetSdk=24
    versionName=1.0
    splits=[base]
    apkSigningVersion=1
    applicationInfo=ApplicationInfo{29cb2a4 com.example.testapp.first}
    flags=[ HAS_CODE ALLOW_CLEAR_USER_DATA ALLOW_BACKUP ]
    privateFlags=[ RESIZEABLE_ACTIVITIES ]
    dataDir=/data/user/0/com.example.testapp.first
    supportsScreens=[small, medium, large, xlarge, resizeable, anyDensity]
    timeStamp=2016-11-03 01:12:08
    firstInstallTime=2016-11-03 01:12:09
    lastUpdateTime=2016-11-03 01:12:09
    signatures=PackageSignatures{9fe380d [53ea108d]}
    installPermissionsFixed=true installStatus=1
    pkgFlags=[ HAS_CODE ALLOW_CLEAR_USER_DATA ALLOW_BACKUP ]
    User 0: ceDataInode=474317 installed=true hidden=false suspended=false stopped=true notLaunched=true enabled=0
      runtime permissions:

``` 




https://stackoverflow.com/questions/40543815/to-what-do-the-android-adb-shell-dumpsys-package-hashes-refer-to

Reply:
So, just to summarize for the sake of history - Jonathan and I discussed this and reached the conclusion that the current implementation indeed is not totally safe, that's why we add the `optimizeAppInstall` flag to enable this optimization in verified contexts (such as CI environment, where the devices are ephemeral and the flow is deterministic).